src/haz3lcore/dynamics/Elaborator.re

```diff
@@ -33,7 +33,11 @@ let cast = (ctx: Ctx.t, mode: Mode.t, self_ty: Typ.t, d: DHExp.t) =>
   | SynFun =>
     switch (self_ty) {
     | Unknown(prov) =>
-      DHExp.cast(d, Unknown(prov), Arrow(Unknown(prov), Unknown(prov)))
+      switch (d) {
+      | Cast(_) => d
```

Comment:
this seems like too general of a fix to handle the specific case of empty holes in function position having casts on them. empty holes should always take on the expected type, and in function position that would be ? -> ?.